File: src/go/generator/lib/instasnowflake/insta_snowflake.go
package instasnowflake

import (
	"fmt"
	"strconv"
	"sync/atomic"
	"time"

	"github.com/dilipkumardk/uuid-generator/src/go/generator/lib"
)

const (
	epoch          = uint64(1767225600000) // Jan 1, 2026
	shardIdBits    = 13
	sequenceBits   = 10
	maxShardId     = (1 << shardIdBits) - 1
	maxSequence    = (1 << sequenceBits) - 1
	shardIdShift   = sequenceBits
	timestampShift = sequenceBits + shardIdBits
)

type InstaSnowflake struct {
	shardId       uint64
	sequence      atomic.Uint64
	lastTimestamp atomic.Uint64
}

func New() *InstaSnowflake {
	return &InstaSnowflake{
		shardId: lib.GetNodeIdFromIp() & maxShardId,
	}
}

func (s *InstaSnowflake) currentTimeMillis() uint64 {
	return uint64(time.Now().UnixMilli())
}

func (s *InstaSnowflake) waitForNextMillis(lastTs uint64) uint64 {
	timestamp := s.currentTimeMillis()
	for timestamp <= lastTs {
		timestamp = s.currentTimeMillis()
	}
	return timestamp
}

func (s *InstaSnowflake) NextId() uint64 {
	timestamp := s.currentTimeMillis()
	lastTs := s.lastTimestamp.Load()

	if timestamp < lastTs {
		fmt.Println("Clock moved backwards. Refusing to generate id.")
		return 0
	}

	if timestamp == lastTs {
		seq := (s.sequence.Add(1)) & maxSequence
		if seq == 0 {
			timestamp = s.waitForNextMillis(lastTs)
		}
	} else {
		s.sequence.Store(0)
	}

	s.lastTimestamp.Store(timestamp)

	id := ((timestamp - epoch) << timestampShift) |
		(s.shardId << shardIdShift) |
		s.sequence.Load()

	return id
}

func (s *InstaSnowflake) NextIdString() string {
	return strconv.FormatUint(s.NextId(), 10)
}

File: src/python/generator/lib/instasnowflake/insta_snowflake.py
import time
import threading
from lib.network_util import get_node_id_from_ip

class InstaSnowflake:
    def __init__(self):
        self.EPOCH = 1767225600000 # Jan 1, 2026
        self.SHARD_ID_BITS = 13
        self.SEQUENCE_BITS = 10
        self.MAX_SHARD_ID = (1 << self.SHARD_ID_BITS) - 1
        self.MAX_SEQUENCE = (1 << self.SEQUENCE_BITS) - 1
        self.SHARD_ID_SHIFT = self.SEQUENCE_BITS
        self.TIMESTAMP_SHIFT = self.SEQUENCE_BITS + self.SHARD_ID_BITS

        self.shard_id = get_node_id_from_ip() & self.MAX_SHARD_ID
        self.sequence = 0
        self.last_timestamp = -1
        self.lock = threading.Lock()

    def current_time_millis(self):
        return int(time.time() * 1000)

    def wait_for_next_millis(self, last_ts):
        timestamp = self.current_time_millis()
        while timestamp <= last_ts:
            timestamp = self.current_time_millis()
        return timestamp

    def next_id_string(self):
        with self.lock:
            timestamp = self.current_time_millis()

            if timestamp < self.last_timestamp:
                print("Clock moved backwards. Refusing to generate id.")
                return "0"

            if timestamp == self.last_timestamp:
                self.sequence = (self.sequence + 1) & self.MAX_SEQUENCE
                if self.sequence == 0:
                    timestamp = self.wait_for_next_millis(self.last_timestamp)
            else:
                self.sequence = 0

            self.last_timestamp = timestamp

            id = ((timestamp - self.EPOCH) << self.TIMESTAMP_SHIFT) | \
                 (self.shard_id << self.SHARD_ID_SHIFT) | \
                 self.sequence

            return str(id)
